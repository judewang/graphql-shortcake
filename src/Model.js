/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
