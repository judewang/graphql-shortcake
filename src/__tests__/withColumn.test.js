import { client } from 'knex';
import _ from 'lodash';
import database from './database';
import Model, {
  Hash, DateTime, Enum, Expiratoin, MixedModel,
} from '..';

class User extends Model {
  static database = database;

  static tableName = 'column_user';

  static columns = {
    name: { type: String },
  }
}

class Member extends Model {
  static database = database;

  static columns = {
    name: { type: String },
  }
}

const UserMixed = new MixedModel([User, Member]);

const SizeEnum = new Enum(['large', 'medium', 'small']);

class Column extends Model {
  static database = database;

  static tableName = 'column';

  static hasOperator = true;

  static columns = {
    aString: { type: String },
    aNumber: { type: Number },
    aBoolean: { type: Boolean },
    aModel: { type: User, name: 'modelId' },
    aMixedModel: { type: UserMixed, name: 'mixedModelId' },
    aHash: { type: Hash },
    aExpiratoin: { type: Expiratoin },
    aDate: { type: DateTime.Date },
    aDatetime: { type: DateTime },
    aEnum: { type: SizeEnum },
    aArray: { type: [String] },
    aArrayModel: { type: [User] },
    aAlias: { type: String, name: 'aliasName' },
    aArchive: { type: String, paths: ['archive'] },
    aArchiveModel: { type: User, paths: ['archive'] },
    aArchiveAlias: { type: String, paths: ['archive'], name: 'aliasName' },
    aOtherPath: { type: Number, paths: ['other'] },
  }
}

describe('Column', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('column');
    await database.schema.createTable('column', (table) => {
      table.increments();
      table.string('name');
      table.string('a_string');
      table.integer('a_number');
      table.boolean('a_boolean');
      table.bigInteger('model_id');
      table.string('mixed_model_id');
      table.text('a_hash');
      table.datetime('a_expiratoin');
      table.date('a_date');
      table.datetime('a_datetime');
      table.specificType('a_enum', 'smallint');
      table.specificType('a_array', 'text[]');
      table.specificType('a_array_model', 'integer[]');
      table.string('alias_name');
      table.jsonb('archive');
      table.jsonb('other');
      table.timestamps();
      table.integer('created_by');
      table.integer('updated_by');
      table.timestamp('deleted_at');
      table.integer('deleted_by');
    });

    await database.schema.dropTableIfExists('column_user');
    await database.schema.createTable('column_user', (table) => {
      table.increments();
      table.string('name');
      table.timestamps();
      table.datetime('deleted_at');
    });

    await new Column().insert({ name: 'banana' });
  });

  it('when value is null', async () => {
    const column = await Column.load(1);
    _.forEach([
      'aString', 'aNumber', 'aBoolean',
      'aModel', 'aMixedModel',
      'aHash', 'aExpiratoin',
      'aDate', 'aDatetime', 'aEnum',
      'aArray', 'aArrayModel', 'aAlias',
      'aArchive', 'aArchiveModel', 'aArchiveAlias',
      'aOtherPath',
    ], (name) => {
      expect(column[name]).toBeNull();
    });
    expect(column).toMatchSnapshot();
  });

  it('when model not found', async () => {
    const column = Column.forge({ archive: { aArchiveModel: '10' } });
    await expect(column.aArchiveModel).resolves.toBe(null);
  });

  it('when column no define', () => {
    const column = new Column();
    expect(column.coffee).toBe(undefined);
  });

  it('save to postgres', async () => {
    const user = new User({ name: 'I`m user' });
    await user.save();

    const model = new Column({
      aString: 'apple',
      aNumber: 99,
      aBoolean: true,
      aModel: user,
      aMixedModel: user,
      aHash: 'mykey',
      aExpiratoin: new Date('2014-03-18T10:00:00'),
      aDate: '2020-01-01',
      aDatetime: new Date('2020-04-01T10:00:00'),
      aEnum: 'medium',
      aArray: ['avocado', 'banana'],
      aArrayModel: [user],
      aAlias: 'cherry',
      aArchive: 'coconut',
      aArchiveModel: user,
      aArchiveAlias: 'grape',
      aOtherPath: 20,
    });

    await model.save(10);
    expect(model).toMatchSnapshot();
    expect(model.id).toBe(Column.toGlobalId('2'));
    expect(`${model.aDate}`).toBe('2020-01-01');
    expect(client).toHaveBeenCalledTimes(2);
  });

  it('fetch from postgres', async () => {
    const model = await Column.load('2');
    expect(model).toMatchSnapshot();
    expect(`${model.aDate}`).toBe('2020-01-01');
    expect(client).toHaveBeenCalledTimes(1);
  });

  describe('columns', () => {
    it('object', () => {
      const { columns } = Column;
      expect(_.keys(columns)).toEqual([
        'aString', 'aNumber', 'aBoolean',
        'aModel', 'aMixedModel', 'aHash',
        'aExpiratoin', 'aDate', 'aDatetime',
        'aEnum', 'aArray', 'aArrayModel',
        'aAlias', 'aArchive', 'aArchiveModel',
        'aArchiveAlias', 'aOtherPath',
        'createdAt', 'updatedAt', 'deletedAt',
        'createdBy', 'updatedBy', 'deletedBy',
      ]);
      expect(columns).toMatchSnapshot();
    });

    it('thunk', () => {
      class Thunk extends Model {
        static database = database;

        static tableName = 'model_table';

        static columns = () => ({
          name: { type: String },
          password: { type: Hash },
          data: { type: Object },
        })
      }
      const { columns } = Thunk;
      expect(_.keys(columns)).toEqual([
        'name', 'password', 'data',
        'createdAt', 'updatedAt', 'deletedAt',
      ]);
      expect(columns).toMatchSnapshot();
    });
  });
});
