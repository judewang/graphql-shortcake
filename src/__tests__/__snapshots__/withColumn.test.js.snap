// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column columns object 1`] = `
Object {
  "aAlias": Object {
    "isArray": false,
    "paths": Array [
      "aliasName",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aArchive": Object {
    "isArray": false,
    "paths": Array [
      "archive",
      "aArchive",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aArchiveAlias": Object {
    "isArray": false,
    "paths": Array [
      "archive",
      "aliasName",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aArchiveModel": Object {
    "isArray": false,
    "paths": Array [
      "archive",
      "aArchiveModel",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aArray": Object {
    "isArray": true,
    "paths": Array [
      "aArray",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aArrayModel": Object {
    "isArray": true,
    "paths": Array [
      "aArrayModel",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aBoolean": Object {
    "isArray": false,
    "paths": Array [
      "aBoolean",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aDate": Object {
    "isArray": false,
    "paths": Array [
      "aDate",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aDatetime": Object {
    "isArray": false,
    "paths": Array [
      "aDatetime",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aEnum": Object {
    "isArray": false,
    "paths": Array [
      "aEnum",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aExpiratoin": Object {
    "isArray": false,
    "paths": Array [
      "aExpiratoin",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aHash": Object {
    "isArray": false,
    "paths": Array [
      "aHash",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aMixedModel": Object {
    "isArray": false,
    "paths": Array [
      "mixedModelId",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aModel": Object {
    "isArray": false,
    "paths": Array [
      "modelId",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aNumber": Object {
    "isArray": false,
    "paths": Array [
      "aNumber",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aOtherPath": Object {
    "isArray": false,
    "paths": Array [
      "other",
      "aOtherPath",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "aString": Object {
    "isArray": false,
    "paths": Array [
      "aString",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "createdAt": Object {
    "isArray": false,
    "paths": Array [
      "createdAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "createdBy": Object {
    "isArray": false,
    "paths": Array [
      "createdBy",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "deletedAt": Object {
    "isArray": false,
    "paths": Array [
      "deletedAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "deletedBy": Object {
    "isArray": false,
    "paths": Array [
      "deletedBy",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "updatedAt": Object {
    "isArray": false,
    "paths": Array [
      "updatedAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "updatedBy": Object {
    "isArray": false,
    "paths": Array [
      "updatedBy",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
}
`;

exports[`Column columns thunk 1`] = `
Object {
  "createdAt": Object {
    "isArray": false,
    "paths": Array [
      "createdAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "data": Object {
    "isArray": false,
    "paths": Array [
      "data",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "deletedAt": Object {
    "isArray": false,
    "paths": Array [
      "deletedAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "name": Object {
    "isArray": false,
    "paths": Array [
      "name",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "password": Object {
    "isArray": false,
    "paths": Array [
      "password",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
  "updatedAt": Object {
    "isArray": false,
    "paths": Array [
      "updatedAt",
    ],
    "toType": [Function],
    "toValue": [Function],
    "type": [Function],
  },
}
`;

exports[`Column fetch from postgres 1`] = `
Column {
  "aAlias": "cherry",
  "aArchive": "coconut",
  "aArchiveAlias": "grape",
  "aArchiveModel": User {
    "then": [Function],
  },
  "aArray": Array [
    "avocado",
    "banana",
  ],
  "aArrayModel": Array [
    User {
      "then": [Function],
    },
  ],
  "aBoolean": true,
  "aDate": "[object DateTime.Date]",
  "aDatetime": "[object DateTime]",
  "aEnum": "medium",
  "aExpiratoin": true,
  "aHash": "[object Hash]",
  "aMixedModel": MixedModel {
    "id": "iN2I0oUUhtxVJpJ8Siv3",
  },
  "aModel": User {
    "then": [Function],
  },
  "aNumber": 99,
  "aOtherPath": 20,
  "aString": "apple",
  "createdAt": "[object DateTime]",
  "createdBy": "10",
  "deletedAt": null,
  "deletedBy": null,
  "id": "iNUoiN3rDinQUv1eySGlvu",
  "updatedAt": "[object DateTime]",
  "updatedBy": "10",
}
`;

exports[`Column save to postgres 1`] = `
Column {
  "aAlias": "cherry",
  "aArchive": "coconut",
  "aArchiveAlias": "grape",
  "aArchiveModel": User {
    "then": [Function],
  },
  "aArray": Array [
    "avocado",
    "banana",
  ],
  "aArrayModel": Array [
    User {
      "then": [Function],
    },
  ],
  "aBoolean": true,
  "aDate": "[object DateTime.Date]",
  "aDatetime": "[object DateTime]",
  "aEnum": "medium",
  "aExpiratoin": true,
  "aHash": "[object Hash]",
  "aMixedModel": MixedModel {
    "id": "iN2I0oUUhtxVJpJ8Siv3",
  },
  "aModel": User {
    "then": [Function],
  },
  "aNumber": 99,
  "aOtherPath": 20,
  "aString": "apple",
  "createdAt": "[object DateTime]",
  "createdBy": "10",
  "deletedAt": null,
  "deletedBy": null,
  "id": "iNUoiN3rDinQUv1eySGlvu",
  "updatedAt": "[object DateTime]",
  "updatedBy": "10",
}
`;

exports[`Column when value is null 1`] = `
Column {
  "aAlias": null,
  "aArchive": null,
  "aArchiveAlias": null,
  "aArchiveModel": null,
  "aArray": null,
  "aArrayModel": null,
  "aBoolean": null,
  "aDate": null,
  "aDatetime": null,
  "aEnum": null,
  "aExpiratoin": null,
  "aHash": null,
  "aMixedModel": null,
  "aModel": null,
  "aNumber": null,
  "aOtherPath": null,
  "aString": null,
  "createdAt": null,
  "createdBy": null,
  "deletedAt": null,
  "deletedBy": null,
  "id": "iNUoiN3rDinQUv1eySGlvt",
  "updatedAt": null,
  "updatedBy": null,
}
`;
